db/src/Pos/DB/Block/Logic/VAR.hs

```diff
@@ -171,16 +178,46 @@ rollingVerifyAndApply
     -> Maybe SlotId
     -> Bool
     -> [NewestFirst NE Blund]
-    -> (OldestFirst NE Block, OldestFirst [] Block)
+    -> [OldestFirst NE Block]
     -> ExceptT ApplyBlocksException m (HeaderHash, NewestFirst [] Blund)
-rollingVerifyAndApply genesisConfig curSlot rollback blunds (prefix, suffix) = do
+rollingVerifyAndApply _ _ _ blunds [] = (,concatNE blunds) <$> lift GS.getTip
+rollingVerifyAndApply genesisConfig curSlot rollback blunds (prefix : suffix) = do
     let prefixHead = prefix ^. _Wrapped . _neHead
-    when (isLeft prefixHead) $ do
-        let epochIndex = prefixHead ^. epochIndexL
-        logDebug $ "Rolling: Calculating LRC if needed for epoch "
-                   <> pretty epochIndex
-        lift $ lrcSingleShot genesisConfig epochIndex
+        epochIndex = prefixHead ^. epochIndexL
+
+    if isLeft prefixHead
+        then do
+            -- This is an EEB for the Original era.
+            logDebug $ "Rolling: Calculating LRC if needed for epoch "
+                    <> pretty epochIndex
+            lift $ lrcSingleShot genesisConfig epochIndex
+        else do
+            -- We may or may not be in the Original era here.
+            -- This is a somewhat savage hack to make sure the adopted block version
+            -- data is updated on the first block of an OBFT epoch (which may not
+            -- be slot 0 of the new epoch).
+            tipHeader <- getTipHeader
+            case tipHeader of
+                BlockHeaderGenesis _ -> pure ()
+
+                BlockHeaderMain mb ->
+                    when (mb ^. epochIndexL == epochIndex - 1) $ do
+                        logDebug $ "Rolling: Calculating OBFT LRC if needed for epoch "
+                                <> pretty epochIndex
+
+                        logDebug $
+                            sformat ("Erik: slot " % build % ", current " % shown % ", other " % shown)
+                                (mb ^. mainHeaderSlot) (getEpochIndex $ mb ^. epochIndexL) (getEpochIndex $ epochIndex - 1)
+
+                        lift $ lrcSingleShot genesisConfig epochIndex
```

Comment:
I think this is "one of the big changes" that fixes the failure-to-adopt-update issue observed on testnet 👍